# Практическая работа №4: Элементы корреляционного анализа. Проверка статистической гипотезы о равенстве коэффициента корреляции нулю.

Выполнили студентки гр. 0382 Деткова Анна и Рубежова Наталия. Вариант №10

## Цель работы

Освоение основных понятий, связанных с корреляционной зависимостью между случайными величинами, статистическими гипотезами и проверкой их «справедливости».

## Основные теоретические положения

Рассматривается система случайных величин $\{X,Y\}$.<br>

Это случайные величины могут быть **независимы**: $f(x,y)=f_1(x) \cdot f_2(y)$.<br>

Иначе может быть:
- **Функциональная зависимость:**<br>
$y=g(x)$.
- **Статистическая зависимость:**<br>
$\phi(x|y) = f(x,y) / f_2(y);\\ \phi(y|x) = f(x,y) / f_1(x)$.<br>
Одним из видов (частным случаем) статистической зависимости является является **корреляционная зависимость**.

---

**Корреляционной** называют статистическую зависимость двух случайных величин, при которой изменение значения одной из СВ приводит к изменению мат. ожидания другой СВ.<br>

$$M(X|y)=q_1(y)\\
M(Y|x)=q_2(x)$$

Такие функции называются **функциями регрессии**.

**Корреляционный момент**: $\mu_{xy} = M\{[x-M(X)] \cdot [y - M(Y)]\}$.<br>

**Коэффициент корреляции**: $r_{xy} = \frac {\mu_{xy}} {\sigma_x \sigma_y}$.<br>

Для коэффициента корреляции справедливо соотношение: $|r_{xy}| \le 1$.<br>

Случайные величины $X$ и $Y$ называются **коррелированными**, если их корреляционный момент (то же самое, что коэффициент корреляции) отличен от нуля. В противном случае эти величины **некоррелированны**.<br>

Если случайные величины $X$ и $Y$, то они **зависимы**. Обратное предположение в общем случае неверно.<br>

Наглядные формы этих утверждений:<br>
- $X$ и $Y$ коррелированы $\Rightarrow$ СВ $X$ и $Y$ зависимы.
- $X$ и $Y$ независимы $\Rightarrow$ СВ $X$ и $Y$ некоррелированы.

Если СВ $X$ и $Y$ коррелированы и обе функции регресии $X$ на $Y$ и $Y$ на $X$ линейны, то говорят, что $X$ и $Y$ связаны **линейной корреляционной зависимостью**. В частности, это имеет место, если двумерная СВ $\{X,Y\}$ распределена нормально.<br>

Коэффициент корреляции служит **мерой тесноты линейной зависимости** между СВ $X$ и $Y$. Если $|r_{xy}| = 1$ зависимость становится **функциональной**.

---

Значение $\overline r_{xy}$ - **статистическая оценка $r_{xy}$** - коэффициента корреляции. Вычисляется по формуле: $\overline r_{xy} = \frac {\sum_{i=1}^{K_y} \sum_{j=1}^{K_x} n_{ij}y_ix_j - N \overline x_{в} \overline y_{в}} {N S_x S_y}$.<br>

Причем: $\sum_{i=1}^{K_y} \sum_{j=1}^{K_x} n_{ij}y_ix_j = \sum_{i=1}^{K_y} y_i \sum_{j=1}^{K_x} n_{ij}x_j = \sum_{j=1}^{K_x} x_j \sum_{i=1}^{K_y} n_{ij}y_i$.

---

**Доверительный интервал для выборочного коэффициента корреляции**<br>

Распределение $\overline r_{xy}$ при определенных условиях можно удовлетворительно аппроксимировать нормальным законом. Однако при увеличении интенсивности корреляционной связи распределение $\overline r_{xy}$ становится все более ассиметричным.<br>

С помощью преобразования Фишера перейдем к СВ $z$:<br>
$$\overline z = 0.5 \ln{\frac {1 + \overline r_{xy}} {1 - \overline r_{xy}}} = 1.1513 \lg{\frac {1 + \overline r_{xy}} {1 - \overline r_{xy}}}$$

Распределение $z$ при неограниченном возрастании объема выборки ассимптотически нормальное со значением СКВО, равным: $$\overline \sigma_z = \frac 1 {\sqrt{N - 3}}$$

В результате доверительный интервал для $r_{xy}$ генеральной совокупности с доверительной вероятностью $\gamma$ определяется по следующей схеме:
1. Вычисляется выборочное значение $\overline z$;
2. Вычисляется значение $\overline \sigma_z$;
3. Доверительный интервал для генерального значения представляется в виде:<br>
$$(\overline z - \lambda(\gamma) \overline \sigma_z, \overline z + \lambda(\gamma) \overline \sigma_z)$$
где значение $\lambda(\gamma)$ должно удовлетворять условию: $\Phi[\lambda(\gamma)]=\frac \gamma 2$.
4. Для пересчета в доверительный интервал для коэффициента корреляции с тем же значением $\gamma$ необходимо воспользоваться обратным преобразованием Фишера:
$$r = th(z) = \frac {e^z - e^{-z}} {e^z + e^{-z}}$$

---

**Проверка гипотезы о значимости выборочного коэффициента корреляции**<br>

Пусть имеется выборка размера $N$ значений двумерной нормально распределенной СВ $\{X,Y\}$ и вычислено значение выборочного коэффициента корреляции $\overline r_{xy} \ne 0$. Поскольку $\overline r_{xy}$ - СВ, то это не значит, что $r_{xy}$ - коэффициент коррреляции для генеральной совокупности тоже отличен от нуля.<br>

Возникает необходимость проверить гипотезу: $$H_0: r_{xy} = 0$$
Альтернативной будет гипотеза: $$H_1: r_{xy} \ne 0$$

Если основная гипотеза $H_0$ отвергается, то это означает, что выборочный коэффициент корреляции $\overline r_{xy}$ значимо отличается от нуля (значим). В противном случае $\overline r_{xy}$ - незначим.<br>

В качестве критерия проверки стат. гипотезы о значимости выборочного коэффициента корреляции можно принять СВ: $$T=\frac {\overline r_{xy} \sqrt{N-2}} {\sqrt{1 - \overline r_{xy}^2}}$$

При справедливости нулевой гипотезы $H_0$ СВ $T$ распределена по закону Стьюдента с $k=N-2$ степенями свободы. Критическая область для данного критерия двусторонняя.<br>

Стандартная схема проверки гипотезы:<br>
1. Вычисляется $T_{набл}$;
2. По заданному значению уровня значимости $\alpha$ и значению $k$ из табл. "Критические точки распределения Стьюдента" определяется значение $t_{крит}(\alpha,k)$;
3. Если $|T_{набл}| \le t_{крит}(\alpha,k)$ - нет оснований отвергать гипотезу $H_0$;
4. Если $|T_{набл}| > t_{крит}(\alpha,k)$ - основная гипотеза $H_0$ с выборочными данными и должна быть отвергнута.

## Постановка задачи

Из заданной генеральной совокупности сформировать выборку по второму признаку. Провести статистическую обработку второй выборки в объёме практических работ №1 и №2, с целью определения точечных статистических оценок параметров распределения исследуемого признака (математического ожидания, дисперсии, среднеквадратичного отклонения, асимметрии, эксцесса и коэффициента вариации). Для системы двух случайных величин $X$ (первый признак) и $Y$ (второй признак) сформировать двумерную выборку и найти статистическую оценку коэффициента корреляции, построить доверительный интервал для коэффициента корреляции и осуществить проверку статистической гипотезы о равенстве коэффициента корреляции нулю. Полученные результаты содержательно проинтерпретировать.

## Выполнение работы

По возможности каждый пункт работы выполняется с помощью кода. Используется язык программирования Python.

In [146]:
# Подключение библиотек
import pandas
import numpy as np
import matplotlib.pyplot as plt
import math

### 1. Провести статистическую обработку второй выборки в объеме практических работ №1 и №2, с целью определения точечных статистических оценок параметров распределения исследуемого признака (математического ожидания, дисперсии, среднеквадратичного отклонения, асимметрии, эксцесса, моды, медианы и коэффициента вариации). Оформить результаты в виде таблицы, сделать выводы.

Загрузка генеральной совокупности и формирование выборки простым бесповторным отбором.

In [189]:
# Загрузка генеральной совокупности из csv-формата в тип данных Data Frame
general_df = pandas.read_csv('sample.csv', comment='#')

# Заданный объем выборки
N = 117

# Формирование выборки. Способ формирования: простой бесповторный отбор
sample_df = general_df.sample(n=N, replace=False)

Построение ранжированного, вариационного и интервального рядов по признакам $\nu$ и $E$.

In [190]:
# Ранжированный ряд по признаку nu
ranked_df_nu = sample_df.sort_values(by='nu')
# Добавление колонки с рангом
ranked_df_nu['rank']=[i for i in range(0, sample_df.shape[0])]
# Ранжированный ряд по признаку E
ranked_df_e = sample_df.sort_values(by='E')
# Добавление колонки с рангом
ranked_df_e['rank']=[i for i in (range(0, sample_df.shape[0]))]

# Вычисление наибольшего и наименьшего значения в выборке по признаку nu
x_max_nu = max(ranked_df_nu['nu'])
x_min_nu = min(ranked_df_nu['nu'])

# Вычисление наибольшего и наименьшего значения в выборке по признаку E
x_max_e = max(ranked_df_e['E'])
x_min_e = min(ranked_df_e['E'])

In [191]:
# Вариационный ряд из ранжированного ряда по признаку nu
variation_df_nu = ranked_df_nu.drop_duplicates(subset='nu', keep='first')

# Вариационный ряд из ранжированного ряда по признаку E
variation_df_e = ranked_df_e.drop_duplicates(subset='E', keep='first')

In [192]:
# Получение интервального ряда
# Объем выборки
N=117
# Количество интервалов по формуле Стерджесса
k = int(np.floor(1 + 3.31 * np.log10(N)))

In [193]:
# Ширина интервалов h_nu
h_nu = int(np.ceil((x_max_nu - x_min_nu) / k))
# Ширина интервалов h_e
h_e = int(np.ceil((x_max_e - x_min_e) / k))

# Интервальный ряд по признаку nu
# Границы интервалов
# Вставим левую границу для i=1 первого интервала
left = x_min_nu + (1-1)*h_nu
left_borders = [left]
right_borders = []
# Вставим в списки итеративно следующие границы
for i in range(1, k):
  left_borders.append(x_min_nu + h_nu * i)
  right_borders.append(x_min_nu + h_nu * i)
# Добавим последнюю правую границу
right = x_min_nu + k*h_nu
right_borders.append(right)
# Разобьем на интервалы
intervals_nu= []
for i in range(len(left_borders)):
  intervals_nu.append((left_borders[i], right_borders[i]))

# Вычисляем середины интервалов
mid_of_intervals_nu = []
for i in range(len(intervals_nu)):
  left, right = intervals_nu[i]
  mid_of_intervals_nu.append((left + right) / 2)

# Абсолютные частоты для интервалов
abs_freq_nu = [0 for i in range(0, k)]
# Считаем частоту попаданий вариант в каждый интервал
for varianta in variation_df_nu['nu']:
  for i in range(len(intervals_nu)):
    left, right = intervals_nu[i]
    if varianta < right and varianta >= left:
      abs_freq_nu[i]+=1
# Количество вариант
variant_count = variation_df_nu['nu'].shape[0]
# Относительные частоты для интервалов
rel_freq_nu = [freq / variant_count for freq in abs_freq_nu]

# Накопленные частоты для интервалов
accum_abs_freq_nu = []
accum_rel_freq_nu = []
for i in range(len(abs_freq_nu)):
  accum_abs_freq_nu.append(np.sum(abs_freq_nu[:i+1]))
  accum_rel_freq_nu.append(np.sum(rel_freq_nu[:i+1]))

# Добавляем необходимые колонки для Таблицы 1
interval_df_nu = pandas.DataFrame({'i': [m for m in range(1, k+1)]})
interval_df_nu["[x_i, x_i+1)"] = intervals_nu
interval_df_nu['Середина i интервала'] = mid_of_intervals_nu
interval_df_nu['n_i'] = abs_freq_nu
interval_df_nu['p*_i'] = rel_freq_nu
interval_df_nu['Накопленная n_i'] = accum_abs_freq_nu
interval_df_nu['Накопленная p*_i'] = accum_rel_freq_nu

print("Интервальный ряд по признаку nu:")
interval_df_nu.style.set_properties(**{'text-align': 'center'}).hide(axis='index')

Интервальный ряд по признаку nu:


In [194]:
# Интервальный ряд по признаку E
# Границы интервалов
# Вставим левую границу для i=1 первого интервала
left = x_min_e + (1-1)*h_e
left_borders = [left]
right_borders = []
# Вставим в списки итеративно следующие границы
for i in range(1, k):
  left_borders.append(x_min_e + h_e * i)
  right_borders.append(x_min_e + h_e * i)
# Добавим последнюю правую границу
right = x_min_e + k*h_e
right_borders.append(right)
# Разобьем на интервалы
intervals_e= []
for i in range(len(left_borders)):
  intervals_e.append((left_borders[i], right_borders[i]))

# Вычисляем середины интервалов
mid_of_intervals_e = []
for i in range(len(intervals_e)):
  left, right = intervals_e[i]
  mid_of_intervals_e.append((left + right) / 2)

# Абсолютные частоты для интервалов
abs_freq_e = [0 for i in range(0, k)]
# Считаем частоту попаданий вариант в каждый интервал
for varianta in variation_df_e['E']:
  for i in range(len(intervals_e)):
    left, right = intervals_e[i]
    if varianta < right and varianta >= left:
      abs_freq_e[i]+=1
# Количество вариант
variant_count = variation_df_e['E'].shape[0]
# Относительные частоты для интервалов
rel_freq_e = [freq / variant_count for freq in abs_freq_e]

# Накопленные частоты для интервалов
accum_abs_freq_e = []
accum_rel_freq_e = []
for i in range(len(abs_freq_e)):
  accum_abs_freq_e.append(np.sum(abs_freq_e[:i+1]))
  accum_rel_freq_e.append(np.sum(rel_freq_e[:i+1]))

# Добавляем необходимые колонки для Таблицы 1
interval_df_e = pandas.DataFrame({'i': [m for m in range(1, k+1)]})
interval_df_e["[x_i, x_i+1)"] = intervals_e
interval_df_e['Середина i интервала'] = mid_of_intervals_e
interval_df_e['n_i'] = abs_freq_e
interval_df_e['p*_i'] = rel_freq_e
interval_df_e['Накопленная n_i'] = accum_abs_freq_e
interval_df_e['Накопленная p*_i'] = accum_rel_freq_e

print("Интервальный ряд по признаку E:")
interval_df_e.style.set_properties(**{'text-align': 'center'}).hide(axis='index')

Интервальный ряд по признаку E:


---

Рассчет условных вариант по признаку $\nu$. А также мат.ожидания, дисперсии и СКО.

In [195]:
# нахождение условного нуля С - середина интервала,
# который находится по середине (округление вверх) интервального ряда - 4 интервал
C_nu = interval_df_nu["Середина i интервала"][round(len(interval_df_nu) / 2)]

# построение таблицы 1
cond_var_df_nu = pandas.DataFrame({'i': [m for m in range(1, k+1)]})
cond_var_df_nu["x"] = mid_of_intervals_nu
cond_var_df_nu["n"] = abs_freq_nu
cond_var_df_nu["u"] = (mid_of_intervals_nu - C_nu) / h_nu
cond_var_df_nu["n*u"] = cond_var_df_nu["n"] * cond_var_df_nu["u"]
cond_var_df_nu["n*u2"] = cond_var_df_nu["n"] * pow(cond_var_df_nu["u"], 2)
cond_var_df_nu["n*u3"] = cond_var_df_nu["n"] * pow(cond_var_df_nu["u"], 3)
cond_var_df_nu["n*u4"] = cond_var_df_nu["n"] * pow(cond_var_df_nu["u"], 4)
cond_var_df_nu["n*(u+1)4"] = cond_var_df_nu["n"] * pow(cond_var_df_nu["u"] + 1, 4)

# рассчет сумм каждого столбца
sums_nu = ["-", cond_var_df_nu["x"].sum(), cond_var_df_nu["n"].sum(), "-",
        cond_var_df_nu["n*u"].sum(), cond_var_df_nu["n*u2"].sum(),
        cond_var_df_nu["n*u3"].sum(), cond_var_df_nu["n*u4"].sum(),
        cond_var_df_nu["n*(u+1)4"].sum()]

# построение таблицы 2
moments_df_nu = pandas.DataFrame({'i': [m for m in range(1, 5)]})
moments_df_nu["nu"] = [sums_nu[4] / N, sums_nu[5] / N, sums_nu[6] / N, sums_nu[7] / N]
moments_df_nu["mu"] = ["-",
                    (moments_df_nu["nu"][1] - pow(moments_df_nu["nu"][0], 2)) * pow(h_nu, 2),
                    (moments_df_nu["nu"][2] - 3 * moments_df_nu["nu"][1] * moments_df_nu["nu"][0] + 2 * pow(moments_df_nu["nu"][0], 3)) * pow(h_nu, 3),
                    (moments_df_nu["nu"][3] - 6 * moments_df_nu["nu"][2] * moments_df_nu["nu"][0] + 6 * moments_df_nu["nu"][1] * pow(moments_df_nu["nu"][0], 2) - 3 * pow(moments_df_nu["nu"][0], 4)) * pow(h_nu, 4)
                    ]

mean_nu_var = moments_df_nu["nu"][0] * h_nu + C_nu
sigma_nu_var2 = moments_df_nu["mu"][1]
sigma_nu_var = np.sqrt(sigma_nu_var2)

Рассчет условных вариант по признаку $E$. А также мат.ожидания, дисперсии и СКО.

In [196]:
# нахождение условного нуля С - середина интервала,
# который находится по середине (округление вверх) интервального ряда - 4 интервал
C_e = interval_df_e["Середина i интервала"][round(len(interval_df_e) / 2)]

# построение таблицы 1
cond_var_df_e = pandas.DataFrame({'i': [m for m in range(1, k+1)]})
cond_var_df_e["x"] = mid_of_intervals_e
cond_var_df_e["n"] = abs_freq_e
cond_var_df_e["u"] = (mid_of_intervals_e - C_e) / h_e
cond_var_df_e["n*u"] = cond_var_df_e["n"] * cond_var_df_e["u"]
cond_var_df_e["n*u2"] = cond_var_df_e["n"] * pow(cond_var_df_e["u"], 2)
cond_var_df_e["n*u3"] = cond_var_df_e["n"] * pow(cond_var_df_e["u"], 3)
cond_var_df_e["n*u4"] = cond_var_df_e["n"] * pow(cond_var_df_e["u"], 4)
cond_var_df_e["n*(u+1)4"] = cond_var_df_e["n"] * pow(cond_var_df_e["u"] + 1, 4)

# рассчет сумм каждого столбца
sums_e = ["-", cond_var_df_e["x"].sum(), cond_var_df_e["n"].sum(), "-",
        cond_var_df_e["n*u"].sum(), cond_var_df_e["n*u2"].sum(),
        cond_var_df_e["n*u3"].sum(), cond_var_df_e["n*u4"].sum(),
        cond_var_df_e["n*(u+1)4"].sum()]

# построение таблицы 2
moments_df_e = pandas.DataFrame({'i': [m for m in range(1, 5)]})
moments_df_e["nu"] = [sums_e[4] / N, sums_e[5] / N, sums_e[6] / N, sums_e[7] / N]
moments_df_e["mu"] = ["-",
                    (moments_df_e["nu"][1] - pow(moments_df_e["nu"][0], 2)) * pow(h_e, 2),
                    (moments_df_e["nu"][2] - 3 * moments_df_e["nu"][1] * moments_df_e["nu"][0] + 2 * pow(moments_df_e["nu"][0], 3)) * pow(h_e, 3),
                    (moments_df_e["nu"][3] - 6 * moments_df_e["nu"][2] * moments_df_e["nu"][0] + 6 * moments_df_e["nu"][1] * pow(moments_df_e["nu"][0], 2) - 3 * pow(moments_df_e["nu"][0], 4)) * pow(h_e, 4)
                    ]

mean_e_var = moments_df_e["nu"][0] * h_e + C_e
sigma_e_var2 = moments_df_e["mu"][1]
sigma_e_var = np.sqrt(sigma_e_var2)

---

Вычисление выборочного среднего по признакам $\nu$ и $E$.

In [197]:
mean_nu = np.sum(sample_df.nu) / N
print(f"Выборочное среднее по признаку nu: {mean_nu}")

mean_e = np.sum(sample_df.E) / N
print(f"Выборочное среднее по признаку E: {mean_e}")

Выборочное среднее по признаку nu: 456.4871794871795
Выборочное среднее по признаку E: 128.27692307692308


---

Вычисление выборочной дисперсии, исправленной выборочной дисперсии и СКО по признакам $\nu$ и $E$.

In [198]:
sigma_nu2 = np.sum((sample_df.nu - mean_nu) ** 2) / N
print(f"Выборочная дисперсия по признаку nu: {sigma_nu2}")

S2_nu = sigma_nu2 * N / (N-1)
S_nu = np.sqrt(S2_nu)
print(f"Исправленная выборочная дисперсия по признаку nu: {S2_nu}")
print(f"Исправленное СКО выборочной дисперсии по признаку nu: {S_nu}")
print()

sigma_e2 = np.sum((sample_df.E - mean_e) ** 2) / N
print(f"Выборочная дисперсия по признаку E: {sigma_e2}")

S2_e = sigma_e2 * N / (N-1)
S_e = np.sqrt(S2_e)
print(f"Исправленная выборочная дисперсия по признаку E: {S2_e}")
print(f"Исправленное СКО выборочной дисперсии по признаку E: {S_e}")

Выборочная дисперсия по признаку nu: 3574.2327416173575
Исправленная выборочная дисперсия по признаку nu: 3605.0450928381965
Исправленное СКО выборочной дисперсии по признаку nu: 60.042027720907264

Выборочная дисперсия по признаку E: 598.6691255752796
Исправленная выборочная дисперсия по признаку E: 603.8300663129975
Исправленное СКО выборочной дисперсии по признаку E: 24.572953959851827


---

 Вычисление статистических оценок коэффициентов асимметрии $a_s^*$ и эксцесса $\epsilon_s^*$ по признакам $\nu$ и $E$.

In [199]:
from scipy.stats import skew, kurtosis

a_s_nu = skew(sample_df.nu, bias=False)
eps_s_nu = kurtosis(sample_df.nu, bias=False)

print(f"Статистическая оценка коэффициента ассиметрии по признаку nu: {a_s_nu}")
print(f"Статистическая оценка коэффициента эксцессы по признаку nu: {eps_s_nu}\n")

a_s_e = skew(sample_df.E, bias=False)
eps_s_e = kurtosis(sample_df.E, bias=False)

print(f"Статистическая оценка коэффициента ассиметрии по признаку E: {a_s_e}")
print(f"Статистическая оценка коэффициента эксцессы по признаку E: {eps_s_e}")

Статистическая оценка коэффициента ассиметрии по признаку nu: -0.041213278892162274
Статистическая оценка коэффициента эксцессы по признаку nu: -0.5488819760442185

Статистическая оценка коэффициента ассиметрии по признаку E: -0.22100019185165928
Статистическая оценка коэффициента эксцессы по признаку E: 0.01798577935603296


---

Вычисление моды $M_o^*$ по признакам $\nu$ и $E$.

In [200]:
#  частота модального интервала
nM = interval_df_nu.max()['n_i']

row_with_nM = interval_df_nu.loc[interval_df_nu['n_i'] == nM]

# нижняя граница модального интервала
x0 = row_with_nM['[x_i, x_i+1)'].to_list()[0][0]
index = row_with_nM['i'].to_list()[0]

nM1 = 0
nM2 = 0
# частоты интервалов до модального и после
if (index == 0) :
    nM2 = interval_df_nu.loc[interval_df_nu['i'] == index + 1]['n_i'].to_list()[0]
elif (index == len(interval_df_nu) - 1) :
    nM1 = interval_df_nu.loc[interval_df_nu['i'] == index - 1]['n_i'].to_list()[0]
else :
    nM2 = interval_df_nu.loc[interval_df_nu['i'] == index + 1]['n_i'].to_list()[0]
    nM1 = interval_df_nu.loc[interval_df_nu['i'] == index - 1]['n_i'].to_list()[0]

M0_nu = x0 + (nM - nM1) / ((nM - nM1) + (nM - nM2)) * h_nu

print(f"Мода по признаку nu: {M0_nu}")

Мода по признаку nu: 433.75


In [202]:
#  частота модального интервала
nM = interval_df_e.max()['n_i']

row_with_nM = interval_df_e.loc[interval_df_e['n_i'] == nM]

# нижняя граница модального интервала
x0 = row_with_nM['[x_i, x_i+1)'].to_list()[0][0]
index = row_with_nM['i'].to_list()[0]

nM1 = 0
nM2 = 0
# частоты интервалов до модального и после
if (index == 0) :
    nM2 = interval_df_e.loc[interval_df_e['i'] == index + 1]['n_i'].to_list()[0]
elif (index == len(interval_df_e) - 1) :
    nM1 = interval_df_e.loc[interval_df_e['i'] == index - 1]['n_i'].to_list()[0]
else :
    nM2 = interval_df_e.loc[interval_df_e['i'] == index + 1]['n_i'].to_list()[0]
    nM1 = interval_df_e.loc[interval_df_e['i'] == index - 1]['n_i'].to_list()[0]

M0_e = x0 + (nM - nM1) / ((nM - nM1) + (nM - nM2)) * h_e

print(f"Мода по признаку E: {M0_e}")

Мода по признаку E: 131.81428571428572


---

Вычисление медианы $M_e^*$ по признаками $\nu$ и $E$.

In [203]:
# медианный интервал
interval_df_less = interval_df_nu.loc[interval_df_nu['Накопленная n_i'] >= round(interval_df_nu.sum()['n_i']/2)]
row_with_frec = interval_df_less.loc[interval_df_less['i'] == len(interval_df_less)]
ind_row_with_frec = row_with_frec['i'].to_list()[0]

# нижняя граница медианного интервала
x02 = row_with_frec['[x_i, x_i+1)'].to_list()[0][0]

# частота медианного интервала
nm = row_with_frec['n_i'].to_list()[0]

# накопленная частота предыдущего интервала
nm2k = 0
if (ind_row_with_frec != 0) :
  nm2k = interval_df_nu.loc[interval_df_nu['i'] == ind_row_with_frec - 1]['Накопленная n_i'].to_list()[0]

Me_nu = x02 + (0.5 * N - nm2k) * h_nu / nm

print(f"Медиана по признаку nu: {Me_nu}")

Медиана по признаку nu: 482.825


In [204]:
# медианный интервал
interval_df_less = interval_df_e.loc[interval_df_e['Накопленная n_i'] >= round(interval_df_e.sum()['n_i']/2)]
row_with_frec = interval_df_less.loc[interval_df_less['i'] == len(interval_df_less)]
ind_row_with_frec = row_with_frec['i'].to_list()[0]

# нижняя граница медианного интервала
x02 = row_with_frec['[x_i, x_i+1)'].to_list()[0][0]

# частота медианного интервала
nm = row_with_frec['n_i'].to_list()[0]

# накопленная частота предыдущего интервала
nm2k = 0
if (ind_row_with_frec != 0) :
  nm2k = interval_df_e.loc[interval_df_e['i'] == ind_row_with_frec - 1]['Накопленная n_i'].to_list()[0]

Me_e = x02 + (0.5 * N - nm2k) * h_e / nm

print(f"Медиана по признаку nu E: {Me_e}")

Медиана по признаку nu E: 130.29642857142858


---

Вычисление коэффициента вариации $V^*$ по признакам $\nu$ и $E$.

In [205]:
# коэффициенты вариации
V_nu = S_nu / mean_nu * 100
V_e = S_e / mean_e * 100

print(f"Коэффициент вариации по признаку nu: {V_nu} %")
print(f"Коэффициент вариации по признаку E: {V_e} %")

Коэффициент вариации по признаку nu: 13.153058928918625 %
Коэффициент вариации по признаку E: 19.15617662977175 %


---

Результаты представлены в виде таблицы:

In [180]:
stats_table = pandas.DataFrame({'Точечная оценка': ["Математическое ожидание", "Дисперсия", "СКО", "Асимметрия", "Эксцесс", "Мода", "Медиана", "Коэффициент вариации"]})
stats_table["nu"] = [mean_nu, sigma_nu2, S_nu, a_s_nu, eps_s_nu, M0_nu, Me_nu, V_nu]
stats_table["E"] = [mean_e, sigma_e2, S_e, a_s_e, eps_s_e, M0_e, Me_e, V_e]
stats_table

Точечная оценка           nu           E
0  Математическое ожидание   453.564103  128.864103
1                Дисперсия  2493.408284  489.286746
2                      СКО    50.148810   22.214966
3               Асимметрия     0.124888   -0.284702
4                  Эксцесс     0.211168    0.279825
5                     Мода   451.181818  125.300000
6                  Медиана   476.780000  130.554545
7     Коэффициент вариации    11.056609   17.239065

**Вывод**: После вычисления точечных оценок можно заподозрить наличие какой-то корреляционной зависимости, но данное предположение не дает подтверждений, поэтом необходим более глубокий анализ.

### 2. Построить двумерный интервальный вариационный ряд, оформить в виде таблицы.

In [206]:
# Интервалы для nu
meanings_in_intervals_nu = {}
for i in intervals_nu :
  meanings_in_intervals_nu[i] = []

for nu in sample_df.nu :
  for interval in intervals_nu :
    left, right = interval
    if nu < right and nu >= left :
      meanings_in_intervals_nu[interval].append(nu)

# Интервалы для E
meanings_in_intervals_e = {}
for i in intervals_e :
  meanings_in_intervals_e[i] = []

for e in sample_df.E :
  for interval in intervals_e :
    left, right = interval
    if e < right and e >= left :
      meanings_in_intervals_e[interval].append(e)

# sum=0
# for k in meanings_in_intervals_e.keys() :
#   sum += len(meanings_in_intervals_e[k])

# print(sum)

In [207]:
two_dim_interval_df = pandas.DataFrame(index=intervals_e, columns=intervals_nu)
two_dim_interval_df.fillna(0, inplace=True)

for i in range(N) :
  for nu_interval in meanings_in_intervals_nu.keys() :
    if sample_df.nu.to_list()[i] in meanings_in_intervals_nu[nu_interval] :
      int_nu = nu_interval
      break
  for e_interval in meanings_in_intervals_e.keys() :
    if sample_df.E.to_list()[i] in meanings_in_intervals_e[e_interval] :
      int_e = e_interval
      break
  two_dim_interval_df[int_nu][int_e] += 1

print("Двумерный интервальный вариацонный ряд:")
two_dim_interval_df.style.set_properties(**{'text-align': 'center'})

Двумерный интервальный вариацонный ряд:


### 3. По полученному двумерному интервальному вариационному ряду построить корреляционную таблицу, сделать выводы.

В качестве значения элемента в интервале будем брать середину каждого интервала при рассчете корреляционной таблицы.

In [208]:
two_dim_int_with_sums_table_df = two_dim_interval_df.copy()
two_dim_int_with_sums_table_df.index = mid_of_intervals_e
two_dim_int_with_sums_table_df.columns = mid_of_intervals_nu

sums_nu = {nu: 0 for nu in mid_of_intervals_nu}
sums_nu['n_E'] = 0
sums_e = {e: 0 for e in mid_of_intervals_e}

for nu in mid_of_intervals_nu :
  for e in mid_of_intervals_e :
    sums_nu[nu] += two_dim_int_with_sums_table_df[nu][e]
    sums_e[e] += two_dim_int_with_sums_table_df[nu][e]

for k in [sums_nu[i] for i in sums_nu.keys()] :
  sums_nu['n_E'] += k

two_dim_int_with_sums_table_df.insert(len(two_dim_int_with_sums_table_df.columns), "n_E", [sums_e[i] for i in sums_e.keys()])
two_dim_int_with_sums_table_df = pandas.concat([two_dim_int_with_sums_table_df,
                         pandas.DataFrame(columns=two_dim_int_with_sums_table_df.columns,
                                          index=['n_nu'])])
for nu in two_dim_int_with_sums_table_df.columns :
  two_dim_int_with_sums_table_df[nu]['n_nu'] = sums_nu[nu]

print("Двумерный интервальный ряд с посчитанными суммами элементов по каждому признаку:")
two_dim_int_with_sums_table_df.style.set_properties(**{'text-align': 'center'})

Двумерный интервальный ряд с посчитанными суммами элементов по каждому признаку:


Как можно заметить $\sum n_{\nu}=\sum n_E=117$ - объем выборки, таблица сумм составлена правильно.

In [210]:
cor_table_df = two_dim_int_with_sums_table_df.copy()
NU_i = {e: 0 for e in mid_of_intervals_e}
NU_i['n_nu'] = '-'
E_i = {nu: 0 for nu in mid_of_intervals_nu}
E_i['n_E'] = '-'
E_i['NU_i'] = '-'
for nu in mid_of_intervals_nu :
  for e in mid_of_intervals_e :
    NU_i[e] += cor_table_df[nu][e] * nu
    E_i[nu] += cor_table_df[nu][e] * e

cor_table_df.insert(len(cor_table_df.columns), "NU_i", [NU_i[i] for i in NU_i.keys()])
cor_table_df = pandas.concat([cor_table_df,
                         pandas.DataFrame(columns=cor_table_df.columns,
                                          index=['E_j'])])
for nu in cor_table_df.columns :
  cor_table_df[nu]['E_j'] = E_i[nu]

NU_i_e = {e: e * NU_i[e] for e in mid_of_intervals_e}
NU_i_e['n_nu'] = '-'
NU_i_e['E_j'] = 0
for e in mid_of_intervals_e :
  NU_i_e['E_j'] += NU_i_e[e]

E_i_nu = {nu: nu * E_i[nu] for nu in mid_of_intervals_nu}
E_i_nu['n_E'] = '-'
E_i_nu['NU_i'] = 0
for nu in mid_of_intervals_nu :
  E_i_nu['NU_i'] += E_i_nu[nu]
E_i_nu['NU_i*e_i'] = '-'

cor_table_df.insert(len(cor_table_df.columns), "NU_i*e_i", [NU_i_e[i] for i in NU_i_e.keys()])
cor_table_df = pandas.concat([cor_table_df,
                         pandas.DataFrame(columns=cor_table_df.columns,
                                          index=['E_j*nu_j'])])

for nu in cor_table_df.columns :
  cor_table_df[nu]['E_j*nu_j'] = E_i_nu[nu]

sum_for_cor = E_i_nu['NU_i']
print(f"Сумма для вычисления коэффицента корреляции = {sum_for_cor}")
print("Корреляционная таблица с вычисленными суммами для вычисления коэффициента корреляции:")
cor_table_df.style.set_properties(**{'text-align': 'center'})

Сумма для вычисления коэффицента корреляции = 7010160.749999999
Корреляционная таблица с вычисленными суммами для вычисления коэффициента корреляции:


Как можно заметить $\sum_{i=1}^{K_y} \sum_{j=1}^{K_x} n_{ij}y_ix_j = \sum_{i=1}^{K_y} y_i \sum_{j=1}^{K_x} n_{ij}x_j = \sum_{j=1}^{K_x} x_j \sum_{i=1}^{K_y} n_{ij}y_i$, в нашем случае $\sum_{j=1}^{K_\nu} \nu_j \sum_{i=1}^{K_E} n_{ij}E_i$ действительно равно $\sum_{i=1}^{K_E} E_i \sum_{j=1}^{K_\nu} n_{ij} \nu_j$, то есть корреляционная таблица составлена верно.

Можно заметить, что значения выборки в основном скоплены вдоль главной диагонали, что позволяет делать предположение о высоком коэффициенте корреляции.

### 4. Исходя из результатов корреляционной таблицы вычислить значение выборочного коэффициента корреляции двумя способами: с помощью стандартной формулы и с помощью условных вариант. Убедиться, что результаты совпадают. Сделать выводы.

Вычисление значения выборочного коэффициента вариации с помощью стандартной формулы и с помощью условных вариант.

In [216]:
r_nu_E = (sum_for_cor - N*mean_nu*mean_e) / (N*S_e*S_nu)
print(f"Выборочный коэффициент корреляции с помощью обычной формулы: {r_nu_E}")
r_nu_E_var = 1 - abs(sum_for_cor - N*mean_nu_var*mean_e_var) / (N*sigma_e_var*sigma_nu_var)
print(f"Выборочный коэффициент корреляции с помощью вариант: {r_nu_E_var}")

Выборочный коэффициент корреляции с помощью обычной формулы: 0.9211900782027463
Выборочный коэффициент корреляции с помощью вариант: 0.8105217253039205


С некой погрешность значения выборочных коэффициентов корреляции совпали, возможно это связано с машинным округлением.

Так как значение выборочного коэффициента корреляции положитель и близко к 1, из чего можно сделать вывод о том, что между случайными величинами наблюдается сильная зависимость, то есть предположение о сильной связи случайных величин подтверждено.

### 5. Построить доверительный интервал для коэффициента корреляции при уровне значимости $\gamma \in \{0,95, 0.99\}$, сделать выводы.

In [222]:
z = 0.5 * math.log((1 + r_nu_E) / (1 - r_nu_E))
print(f"Случайная величина z = {z}")

sigma_z = 1 / math.sqrt(N-3)
print(f"СКВО для распределения СВ z: sigma_z = {sigma_z}")

gamma095 = 0.95
z_conf_interval_gamma_095 = (z - gamma095 / 2 * sigma_z, z + gamma095 / 2 * sigma_z)
print(f"Доверительный интервал СВ z при уровне значимости gamma = 0.95: {z_conf_interval_gamma_095}")
r_conf_interval_gamma_095 = ((math.e ** z_conf_interval_gamma_095[0] - math.e ** (-z_conf_interval_gamma_095[0])) / (math.e ** z_conf_interval_gamma_095[0] + math.e ** (-z_conf_interval_gamma_095[0])),
                             (math.e ** z_conf_interval_gamma_095[1] - math.e ** (-z_conf_interval_gamma_095[1])) / (math.e ** z_conf_interval_gamma_095[1] + math.e ** (-z_conf_interval_gamma_095[1])))
print(f"Доверительный интервал коэффициента корреляции при уровне значимости gamma = 0.95: {r_conf_interval_gamma_095}")

gamma099 = 0.99
z_conf_interval_gamma_099 = (z - gamma099 / 2 * sigma_z, z + gamma099 / 2 * sigma_z)
print(f"Доверительный интервал  СВ z при уровне значимости gamma = 0.99: {z_conf_interval_gamma_099}")
r_conf_interval_gamma_099 = ((math.e ** z_conf_interval_gamma_099[0] - math.e ** (-z_conf_interval_gamma_099[0])) / (math.e ** z_conf_interval_gamma_099[0] + math.e ** (-z_conf_interval_gamma_099[0])),
                             (math.e ** z_conf_interval_gamma_099[1] - math.e ** (-z_conf_interval_gamma_099[1])) / (math.e ** z_conf_interval_gamma_099[1] + math.e ** (-z_conf_interval_gamma_099[1])))
print(f"Доверительный интервал коэффициента корреляции при уровне значимости gamma = 0.99: {r_conf_interval_gamma_099}")


Случайная величина z = 1.5968306026659895
СКВО для распределения СВ z: sigma_z = 0.0936585811581694
Доверительный интервал СВ z при уровне значимости gamma = 0.95: (1.552342776615859, 1.64131842871612)
Доверительный интервал коэффициента корреляции при уровне значимости gamma = 0.95: (0.9141712126126045, 0.9276566513086422)
Доверительный интервал  СВ z при уровне значимости gamma = 0.99: (1.5504696049926956, 1.6431916003392835)
Доверительный интервал коэффициента корреляции при уровне значимости gamma = 0.99: (0.9138629398590661, 0.9279174175400923)


При увеличении уровня значимости расширяется доверительный интервал, а также можно увидеть, что доверительные интервалы показывают, что значение коэффииента корреляции довльно велико.

### 6. Осуществить проверку статистической гипотезы о равенстве коэффициента корреляции нулю при заданном уровне значимости $\alpha = 0.05$, сделать выводы.

Сформулируем гипотезы:
$$H_0 : r_{xy} = 0$$
$$H_1 : r_{xy} \neq 0$$

Осуществим проверку статистической гипотезы о равенстве коэффициента корреляции нулю.

In [224]:
T_obs = (r_nu_E * math.sqrt(N - 2) / math.sqrt(1 - r_nu_E ** 2))
print(f"Т_набл = {T_obs}")
df = N - 2
print(f"При справедливости нулевой гипотезы СВ Т распределена по закону Стьюдента с k = {df} степенями свободы.")

Т_набл = 25.387653135667637
При справедливости нулевой гипотезы СВ Т распределена по закону Стьюдента с k = 115 степенями свободы.


$T_{набл} = 25.39, k = 115$, критическая область для данного критерия двусторонняя.

Найдем по таблице $t_{крит}(\alpha,k) = 1.98$.

$|T_{набл}| > t_{крит}(\alpha,k) (25.39 > 1.98)$, поэтому нулевая гипотеза $H_0$ отвергается.

Это означает, что коэффициент корреляции значимо отличается от нуля, можно говорить о том, что между СВ $\nu$ и $E$ наблюдается корреляционная зависимость при уровне значимости $\alpha = 0.05$.



## Выводы

В ходе выполнения практической работы была проведена статистическая обработка второй выборки в объеме работ №1 и №2 с целью вычисления точечных статистических оценок параметров распределения исследуемого признака. В результате можно было предположить, что некоторая корреляционная зависимость возможна.

Далее был создан двумерный интервальный вариационный ряд и на его основе составлена корреляционная таблица, с использованием которой была рассчитана выборочная оценка коэффициента корреляции.

Кроме того были вычислены доверительные интервалы для данного коэффициента и проверена гипотеза. Результаты показали, что коэффициент корреляции достоверно отличается от нуля, подтверждая наличие линейной связи между выборками.